File: rustbridge-java/rustbridge-ffm/src/main/java/com/rustbridge/ffm/NativeBindings.java
package com.rustbridge.ffm;

import java.lang.foreign.*;
import java.lang.invoke.MethodHandle;

/**
 * Native function bindings for rustbridge FFI.
 * <p>
 * This class holds the method handles for calling native plugin functions.
 */
public class NativeBindings {
    /**
     * Memory layout for RbResponse struct (binary transport response).
     * <pre>
     * struct RbResponse {
     *     error_code: u32,  // 0 = success
     *     len: u32,         // response data size
     *     capacity: u32,    // allocation capacity
     *     _padding: u32,    // alignment padding
     *     data: *mut c_void // response data pointer
     * }
     * </pre>
     */
    public static final StructLayout RB_RESPONSE_LAYOUT = MemoryLayout.structLayout(
            ValueLayout.JAVA_INT.withName("error_code"),
            ValueLayout.JAVA_INT.withName("len"),
            ValueLayout.JAVA_INT.withName("capacity"),
            MemoryLayout.paddingLayout(4),  // Alignment for pointer
            ValueLayout.ADDRESS.withName("data")
    );

    private final MethodHandle pluginInit;
    private final MethodHandle pluginCall;
    private final MethodHandle pluginCallRaw;      // nullable - binary transport optional
    private final MethodHandle pluginFreeBuffer;
    private final MethodHandle rbResponseFree;     // nullable - binary transport optional
    private final MethodHandle pluginShutdown;
    private final MethodHandle pluginSetLogLevel;
    private final MethodHandle pluginGetState;
    private final MethodHandle pluginGetRejectedCount;
    private final boolean hasBinaryTransport;

    /**
     * Create bindings from a symbol lookup.
     *
     * @param lookup the symbol lookup for the native library
     * @param linker the native linker
     */
    public NativeBindings(SymbolLookup lookup, Linker linker) {
        // FfiBuffer layout (for return type)
        // struct FfiBuffer { data: *mut u8, len: usize, capacity: usize, error_code: u32 }
        StructLayout ffiBufferLayout = MemoryLayout.structLayout(
                ValueLayout.ADDRESS.withName("data"),
                ValueLayout.JAVA_LONG.withName("len"),
                ValueLayout.JAVA_LONG.withName("capacity"),
                ValueLayout.JAVA_INT.withName("error_code"),
                MemoryLayout.paddingLayout(4) // Alignment padding
        );

        // plugin_init(plugin_ptr, config_json, config_len, log_callback) -> handle
        this.pluginInit = linker.downcallHandle(
                lookup.find("plugin_init").orElseThrow(),
                FunctionDescriptor.of(
                        ValueLayout.ADDRESS,  // return: handle
                        ValueLayout.ADDRESS,  // plugin_ptr
                        ValueLayout.ADDRESS,  // config_json
                        ValueLayout.JAVA_LONG, // config_len
                        ValueLayout.ADDRESS   // log_callback (nullable)
                )
        );

        // plugin_call(handle, type_tag, request, request_len) -> FfiBuffer
        this.pluginCall = linker.downcallHandle(
                lookup.find("plugin_call").orElseThrow(),
                FunctionDescriptor.of(
                        ffiBufferLayout,      // return: FfiBuffer
                        ValueLayout.ADDRESS,  // handle
                        ValueLayout.ADDRESS,  // type_tag
                        ValueLayout.ADDRESS,  // request
                        ValueLayout.JAVA_LONG // request_len
                )
        );

        // plugin_call_raw(handle, message_id, request, request_size) -> RbResponse
        // Optional - binary transport may not be available
        var callRawSymbol = lookup.find("plugin_call_raw");
        if (callRawSymbol.isPresent()) {
            this.pluginCallRaw = linker.downcallHandle(
                    callRawSymbol.get(),
                    FunctionDescriptor.of(
                            RB_RESPONSE_LAYOUT,   // return: RbResponse
                            ValueLayout.ADDRESS,  // handle
                            ValueLayout.JAVA_INT, // message_id
                            ValueLayout.ADDRESS,  // request
                            ValueLayout.JAVA_LONG // request_size
                    )
            );
        } else {
            this.pluginCallRaw = null;
        }

        // plugin_free_buffer(buffer)
        this.pluginFreeBuffer = linker.downcallHandle(
                lookup.find("plugin_free_buffer").orElseThrow(),
                FunctionDescriptor.ofVoid(
                        ValueLayout.ADDRESS   // buffer pointer
                )
        );

        // rb_response_free(response)
        // Optional - binary transport may not be available
        var rbResponseFreeSymbol = lookup.find("rb_response_free");
        if (rbResponseFreeSymbol.isPresent()) {
            this.rbResponseFree = linker.downcallHandle(
                    rbResponseFreeSymbol.get(),
                    FunctionDescriptor.ofVoid(
                            ValueLayout.ADDRESS   // response pointer
                    )
            );
        } else {
            this.rbResponseFree = null;
        }

        // Binary transport is available if both symbols are present
        this.hasBinaryTransport = this.pluginCallRaw != null && this.rbResponseFree != null;

        // plugin_shutdown(handle) -> bool
        this.pluginShutdown = linker.downcallHandle(
                lookup.find("plugin_shutdown").orElseThrow(),
                FunctionDescriptor.of(
                        ValueLayout.JAVA_BOOLEAN, // return: success
                        ValueLayout.ADDRESS       // handle
                )
        );

        // plugin_set_log_level(handle, level)
        this.pluginSetLogLevel = linker.downcallHandle(
                lookup.find("plugin_set_log_level").orElseThrow(),
                FunctionDescriptor.ofVoid(
                        ValueLayout.ADDRESS,  // handle
                        ValueLayout.JAVA_BYTE // level
                )
        );

        // plugin_get_state(handle) -> u8
        this.pluginGetState = linker.downcallHandle(
                lookup.find("plugin_get_state").orElseThrow(),
                FunctionDescriptor.of(
                        ValueLayout.JAVA_BYTE, // return: state code
                        ValueLayout.ADDRESS    // handle
                )
        );

        // plugin_get_rejected_count(handle) -> u64
        this.pluginGetRejectedCount = linker.downcallHandle(
                lookup.find("plugin_get_rejected_count").orElseThrow(),
                FunctionDescriptor.of(
                        ValueLayout.JAVA_LONG, // return: rejected count
                        ValueLayout.ADDRESS    // handle
                )
        );
    }

    public MethodHandle pluginInit() {
        return pluginInit;
    }

    public MethodHandle pluginCall() {
        return pluginCall;
    }

    public MethodHandle pluginCallRaw() {
        return pluginCallRaw;
    }

    public MethodHandle pluginFreeBuffer() {
        return pluginFreeBuffer;
    }

    public MethodHandle rbResponseFree() {
        return rbResponseFree;
    }

    public MethodHandle pluginShutdown() {
        return pluginShutdown;
    }

    public MethodHandle pluginSetLogLevel() {
        return pluginSetLogLevel;
    }

    public MethodHandle pluginGetState() {
        return pluginGetState;
    }

    public MethodHandle pluginGetRejectedCount() {
        return pluginGetRejectedCount;
    }

    /**
     * Check if binary transport is supported by this plugin.
     *
     * @return true if binary transport is available
     */
    public boolean hasBinaryTransport() {
        return hasBinaryTransport;
    }
}

File: rustbridge-csharp/RustBridge.Native/NativeBindings.cs
using System.Runtime.InteropServices;

namespace RustBridge.Native;

/// <summary>
/// Native function bindings for RustBridge FFI.
/// <para>
/// This class provides P/Invoke declarations for calling native plugin functions.
/// The actual library is loaded dynamically via <see cref="NativePluginLoader"/>.
/// </para>
/// </summary>
internal static class NativeBindings
{
    /// <summary>
    /// FfiBuffer structure returned by plugin_call.
    /// <code>
    /// struct FfiBuffer {
    ///     data: *mut u8,      // pointer to data
    ///     len: usize,         // length of data
    ///     capacity: usize,    // allocation capacity
    ///     error_code: u32     // 0 = success
    /// }
    /// </code>
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct FfiBuffer
    {
        public IntPtr Data;
        public nuint Len;
        public nuint Capacity;
        public uint ErrorCode;
    }

    /// <summary>
    /// RbResponse structure returned by plugin_call_raw (binary transport).
    /// <code>
    /// struct RbResponse {
    ///     error_code: u32,  // 0 = success
    ///     len: u32,         // response data size
    ///     capacity: u32,    // allocation capacity
    ///     _padding: u32,    // alignment padding
    ///     data: *mut c_void // response data pointer
    /// }
    /// </code>
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct RbResponse
    {
        public uint ErrorCode;
        public uint Len;
        public uint Capacity;
        private readonly uint _padding;
        public IntPtr Data;
    }

    /// <summary>
    /// Delegate type for the log callback function.
    /// </summary>
    /// <param name="level">Log level (0=Trace, 1=Debug, 2=Info, 3=Warn, 4=Error).</param>
    /// <param name="target">Pointer to null-terminated target string.</param>
    /// <param name="message">Pointer to message bytes.</param>
    /// <param name="messageLen">Length of the message.</param>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate void LogCallbackDelegate(byte level, IntPtr target, IntPtr message, nuint messageLen);

    /// <summary>
    /// Create a new plugin instance.
    /// </summary>
    /// <returns>Pointer to the plugin instance.</returns>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate IntPtr PluginCreateDelegate();

    /// <summary>
    /// Initialize a plugin with configuration.
    /// </summary>
    /// <param name="pluginPtr">Pointer from plugin_create.</param>
    /// <param name="configJson">Pointer to config JSON bytes.</param>
    /// <param name="configLen">Length of config JSON.</param>
    /// <param name="logCallback">Optional log callback (can be null).</param>
    /// <returns>Handle to the initialized plugin, or null on failure.</returns>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate IntPtr PluginInitDelegate(IntPtr pluginPtr, IntPtr configJson, nuint configLen, IntPtr logCallback);

    /// <summary>
    /// Call the plugin with a JSON request.
    /// </summary>
    /// <param name="handle">Plugin handle from plugin_init.</param>
    /// <param name="typeTag">Null-terminated type tag string.</param>
    /// <param name="request">Pointer to request bytes.</param>
    /// <param name="requestLen">Length of request.</param>
    /// <returns>FfiBuffer containing the response.</returns>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate FfiBuffer PluginCallDelegate(IntPtr handle, IntPtr typeTag, IntPtr request, nuint requestLen);

    /// <summary>
    /// Call the plugin with a binary request (raw transport).
    /// </summary>
    /// <param name="handle">Plugin handle from plugin_init.</param>
    /// <param name="messageId">Binary message ID.</param>
    /// <param name="request">Pointer to request struct.</param>
    /// <param name="requestSize">Size of request struct.</param>
    /// <returns>RbResponse containing the response.</returns>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate RbResponse PluginCallRawDelegate(IntPtr handle, int messageId, IntPtr request, nuint requestSize);

    /// <summary>
    /// Free a buffer returned by plugin_call.
    /// </summary>
    /// <param name="buffer">Pointer to the FfiBuffer.</param>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate void PluginFreeBufferDelegate(IntPtr buffer);

    /// <summary>
    /// Free a response returned by plugin_call_raw.
    /// </summary>
    /// <param name="response">Pointer to the RbResponse.</param>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate void RbResponseFreeDelegate(IntPtr response);

    /// <summary>
    /// Shutdown the plugin.
    /// </summary>
    /// <param name="handle">Plugin handle.</param>
    /// <returns>True if shutdown was successful.</returns>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    [return: MarshalAs(UnmanagedType.U1)]
    public delegate bool PluginShutdownDelegate(IntPtr handle);

    /// <summary>
    /// Set the log level.
    /// </summary>
    /// <param name="handle">Plugin handle.</param>
    /// <param name="level">Log level code.</param>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate void PluginSetLogLevelDelegate(IntPtr handle, byte level);

    /// <summary>
    /// Get the plugin state.
    /// </summary>
    /// <param name="handle">Plugin handle.</param>
    /// <returns>State code (0-5), or 255 for invalid handle.</returns>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate byte PluginGetStateDelegate(IntPtr handle);

    /// <summary>
    /// Get the count of rejected requests.
    /// </summary>
    /// <param name="handle">Plugin handle.</param>
    /// <returns>Number of rejected requests.</returns>
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate ulong PluginGetRejectedCountDelegate(IntPtr handle);
}
